### PR TITLE
[connectors-sdk] Create `ConnectorStateManager` class

### LINKED ISSUE #6206: [Connectors-SDK] Implement StateManager
## Use case

Add StateManager base class to connectors-sdk to standardize connector state
- parsing
- validation
- serialization
-  eliminate duplicated state management logic across connectors.

## Current Workaround

Manually implemented in each connector

## Proposed Solution

Use pydantic base model (as done for base connector settings)
`StateManager[State]`
    - `load() -> State`
    - `save(state: State) -> None`
    - `migrate(raw_state: dict) -> dict`

### PR DESCRIPTION
**Pull request by Powlinett**
### Proposed changes

* create `ConnectorStateManager` class (inherits from `pydantic.BaseModel`)
* create unit tests
* add TDR

### Related issues

* close #6206

### Checklist

- [x] I consider the submitted work as finished
- [x] I have signed my commits using GPG key.
- [x] I tested the code for its functionality using different use cases
- [x] I added/update the relevant documentation (either on github or on notion) -> TDR
- [x] ~~Where necessary I refactored code to improve the overall quality~~

### Further comments

- I chose to use `helper.force_ping()` to update state immediately on OpenCTI, although it could be removed and the state will be updated on connector's heartbeats (approximately every 40 seconds).
- I kept the semantic of "state manager" to be compliant with the issue, but it might be confused with the concept of "state machines". Actually, `ConnectorStateManager` class could be called `ConnectorState` as it only defines the attributes of an object and some CRUD methods (read/update).

### CODE REVIEW

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+        super().__init__()
+
+        self.helper = helper
+
+    def load(self) -> None:
+        """Overwrite instance's fields with the connector's state stored on OpenCTI."""
+        state = self.helper.get_state() or {}
+        for key in state:
+            setattr(self, key, state[key])
+
```
[the remark this answers is not in the document]
Comment (Powlinett) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py:76]: Extra fields must be handled both during `load` and `save`. I commit a fix.

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    def save(self) -> None:
+        """Save instance's fields as connector's state on OpenCTI."""
+        declared_fields = set(type(self).model_fields)
+        state_dict = self.model_dump(mode="json", include=declared_fields)
+
+        self.helper.set_state(state_dict)
+        self.helper.force_ping()  # ensure the state is updated immediately on OpenCTI
```
[the remark this answers is not in the document]
Comment (Powlinett) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py]: Let's use the same serializer in both modules (i.e., `datetime.isoformat()`).

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        """Replace the default JSON serializer, in order to use +00:00 offset instead of Z prefix.
+        This is convenient so `assert self.model_dump(mode="json")["last_run"] == self.last_run.isoformat()`
+        is `True` across both codebase and tests (using the same serializer).
+        Consistent with `DatetimeFromIsoString` in `connectors_sdk.settings.annotated_types` module too.
+        """
+        if isinstance(value, datetime):
+            return value.isoformat()  # Override default JSON serializer
+        return handler(value)
+
+    def load(self) -> None:
```
Comment (throuxel) [inline connectors-sdk/connectors_sdk/state_manager/state_manager.py]: I would say `load` can be automatically called during the `__init__`

Comment (throuxel) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py]: Probably not

Comment (throuxel) [inline connectors-sdk/connectors_sdk/state_manager/state_manager.py]: We can have an flag saying state have been already loaded to prevent desynchronization if we try to load again before saving.

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            return value.isoformat()  # Override default JSON serializer
+        return handler(value)
+
+    def load(self, force: bool = False) -> None:
+        """Overwrite instance's fields with the connector's state stored on OpenCTI.
+
+        Arguments:
+            force: If `True`, load the state from OpenCTI even if there are unsaved changes in the state manager instance.
+        """
+        if self._sync or force:
```
Comment (throuxel) [inline connectors-sdk/connectors_sdk/state_manager/state_manager.py]:
```suggestion
        if not self._sync or force:
```

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        state_dict = self.model_dump(mode="json", include=declared_fields)
+        # Send both declared _and_ extra fields to not delete any connector state's attributes on OpenCTI
+        if self.model_extra:
+            state_dict.update(self.model_extra)
+
+        self._helper.set_state(state_dict)
+        # Ensure the state is updated immediately on OpenCTI (instead of waiting for the next ping)
+        self._helper.force_ping()
+        # OpenCTI is now synchronized with the state manager instance
+        self._sync = True
```
Comment (throuxel) [inline connectors-sdk/connectors_sdk/state_manager/state_manager.py]:
```suggestion
        self._sync = False
```
If load put it to True, save should put it to False

Comment (Powlinett) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py]: Maybe my intent is not very clear and I should rename `_sync` with the full length adjective `_synchronized`.

The intent is to track whether the instance of `ConnectorStateManager` is synchronized with the state stored on OpenCTI.

So:
- During `load` call, the fields of `ConnectorStateManager` are overwritten with the ones stored on OpenCTI. At the end of the function, both states are equals: they're synchronized.
- During `save` call, the fields stored on OpenCTI are overwritten with the fields of `ConnectorStateManager` At the end of the function, both states are equals: they're synchronized.
- When a field of `ConnectorStateManager` is (re)assigned e.g., `state.last_run = datetime(2026, 27, 4, tzinfo=timezone.utc)`, `ConnectorStateManager` and the state stored on OpenCTI are not equals anymore, they're not synchronized (until `save` is called).

For more accuracy, we could check if they're synchronized by comparing the two dicts before calling `setattr` on each `ConnectorStateManager`'s fields. Actually it could make the code easier to understand too. WDYT?

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py (hunk 4 above)
Comment (Powlinett) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py]: This commit has been reverted

File: connectors-sdk/connectors_sdk/state_manager/state_manager.py (hunk 5 above)
Comment (Powlinett) [reply connectors-sdk/connectors_sdk/state_manager/state_manager.py]: This commit has been reverted

**Review by throuxel — APPROVED** [review]
Image

**Review by Ninoxe — APPROVED** [review]
